### PR TITLE
[FIX] account_payment: allow internal users to change a payment provider

### PR DESCRIPTION
**Pull request by barracudapps**
**Issue:**
An employee can't modify a `Payment Method` (other than "Manual" on a new payment entry).

**Expected:**
An employee should be able to change the payment provider to register a customer payment.

**Steps to reproduce:**
- Activate Accounting app and go to Configuration > Online Payments > Payment Providers;
- Install any of these providers, configure it, activate the test mode and publish it;
- Go to Customers Payments > Payments;
- Open or create a payement record;
- Try to change the Payment Method field for .

**Cause:**
No rights have been given to users on the payment provider data lookup.
https://github.com/odoo/odoo/blob/18.0/addons/payment/security/ir.model.access.csv

**Fix:**
Give temporary rights on payment provider token lookup.

opw-4270781

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by nle-odoo — APPROVED** [review]
I would assume that it is a very important model that would have its security implemented as intended. Employee had access to it before 2021 but the access was removed with 573ed74c121c1572b3fab6f9553ed7f93f7b3f99 with this reasoning in task-2085989:

> ### Security
> Only give access to acquirers for system users. Add sudo() in controllers wherever necessary.

(payment.acquirer is the previous name of payment.provider)

So I think the scenario should work without adding access right, checking this particular case the issue is that we do:

https://github.com/odoo/odoo/blob/bb40cf9efc87929e16671538aa68b191dffd06b3/addons/account_payment/models/account_payment.py#L117

that will do a search on payment.provider without access right. So to add sudo we could just add sudo on this whole search, or do a subquery in sudo or we could even just check  `capture_manually` separately:

```diff
--- a/addons/account_payment/models/account_payment.py
+++ b/addons/account_payment/models/account_payment.py
@@ -111,10 +111,13 @@ class AccountPayment(models.Model):
             self.payment_token_id = False
             return
 
+        if self.payment_method_line_id.payment_provider_id.sudo().capture_manually != False:
+            self.payment_token_id = False
+            return
+
         self.payment_token_id = self.env['payment.token'].search([
             *self.env['payment.token']._check_company_domain(self.company_id),
             ('partner_id', '=', self.partner_id.id),
-            ('provider_id.capture_manually', '=', False),
             ('provider_id', '=', self.payment_method_line_id.payment_provider_id.id),
          ], limit=1)
```
Any of these solution would look good to me.

**Comment by barracudapps** [comment]
@nle-odoo Thank you for the more precise investigation and these solutions !
I will indeed give temporary right on token lookup. That way, we should avoid future issues as the lookup might become more complex.

File: addons/account_payment/models/account_payment.py [hunk 1]
```diff
@@ -111,7 +111,7 @@ def _onchange_set_payment_token_id(self):
             self.payment_token_id = False
             return
 
-        self.payment_token_id = self.env['payment.token'].search([
+        self.payment_token_id = self.env['payment.token'].sudo().search([
```
Comment (AntoineVDV) [inline addons/account_payment/models/account_payment.py:114]: I recommend always adding a comment to justify the use of `sudo()`; it makes further maintenance and debugging way easier.
```py
self.payment_token_id = self.env['payment.token'].sudo().search([
            *self.env['payment.token']._check_company_domain(self.company_id),
            ('partner_id', '=', self.partner_id.id),
            ('provider_id.capture_manually', '=', False),
            ('provider_id', '=', self.payment_method_line_id.payment_provider_id.id),
         ], limit=1)  # In sudo mode to read the provider fields.
```

**Review by AntoineVDV — APPROVED** [review]
LGTM, thanks :)

@robodoo r+

**Pull request closed without merging**